### PR TITLE
[Arrow] Fix issue surrounding lifetime of dictionary arrays

### LINKED ISSUE #10608: Crash while scanning PyArrow Dataset
### What happens?

When trying to scan a large number of Feather files via `pyarrow.dataset` I get a segfault in `duckdb::StringHeap::AddBlob`:

```
#0  0x00007ffff6f27e85 in __memmove_avx_unaligned_erms () from /usr/lib64/libc.so.6
#1  0x00007fffd6401c26 in duckdb::StringHeap::AddBlob(char const*, unsigned long) ()
   from /home/maubury/.conda/envs/juggernaut/lib/python3.11/site-packages/duckdb/duckdb.cpython-311-x86_64-linux-gnu.so
#2  0x00007fffd644b997 in void duckdb::ColumnDataCopy<duckdb::string_t>(duckdb::ColumnDataMetaData&, duckdb::UnifiedVectorFormat const&, duckdb::Vector&, unsigned long, unsigned long)
    () from /home/maubury/.conda/envs/juggernaut/lib/python3.11/site-packages/duckdb/duckdb.cpython-311-x86_64-linux-gnu.so
#3  0x00007fffd644df4a in duckdb::ColumnDataCollection::Append(duckdb::ColumnDataAppendState&, duckdb::DataChunk&) ()
   from /home/maubury/.conda/envs/juggernaut/lib/python3.11/site-packages/duckdb/duckdb.cpython-311-x86_64-linux-gnu.so
#4  0x00007fffd697d415 in duckdb::PhysicalBatchCollector::Sink(duckdb::ExecutionContext&, duckdb::DataChunk&, duckdb::OperatorSinkInput&) const ()
   from /home/maubury/.conda/envs/juggernaut/lib/python3.11/site-packages/duckdb/duckdb.cpython-311-x86_64-linux-gnu.so
#5  0x00007fffd6d7ef6a in duckdb::PipelineExecutor::ExecutePushInternal(duckdb::DataChunk&, unsigned long) ()
   from /home/maubury/.conda/envs/juggernaut/lib/python3.11/site-packages/duckdb/duckdb.cpython-311-x86_64-linux-gnu.so
#6  0x00007fffd6d834e5 in duckdb::PipelineExecutor::Execute(unsigned long) ()
   from /home/maubury/.conda/envs/juggernaut/lib/python3.11/site-packages/duckdb/duckdb.cpython-311-x86_64-linux-gnu.so
#7  0x00007fffd6d88529 in duckdb::PipelineTask::ExecuteTask(duckdb::TaskExecutionMode) ()
   from /home/maubury/.conda/envs/juggernaut/lib/python3.11/site-packages/duckdb/duckdb.cpython-311-x86_64-linux-gnu.so
#8  0x00007fffd6d794f3 in duckdb::ExecutorTask::Execute(duckdb::TaskExecutionMode) ()
   from /home/maubury/.conda/envs/juggernaut/lib/python3.11/site-packages/duckdb/duckdb.cpython-311-x86_64-linux-gnu.so
#9  0x00007fffd6d7ae85 in duckdb::TaskScheduler::ExecuteForever(std::atomic<bool>*) ()
   from /home/maubury/.conda/envs/juggernaut/lib/python3.11/site-packages/duckdb/duckdb.cpython-311-x86_64-linux-gnu.so
#10 0x00007fffd70b29e0 in execute_native_thread_routine () from /home/maubury/.conda/envs/juggernaut/lib/python3.11/site-packages/duckdb/duckdb.cpython-311-x86_64-linux-gnu.so
#11 0x00007ffff7bb71cf in start_thread () from /usr/lib64/libpthread.so.0
#12 0x00007ffff6e91dd3 in clone () from /usr/lib64/libc.so.6
```

### To Reproduce

The code I'm executing is as follows:
```
import duckdb
import pyarrow.dataset as ds

files = [...]
duckdb.arrow(ds.dataset(files, format="feather")).arrow()
```
Unfortunately I'm unable to provide the files which trigger the crash :-(

I can confirm that all files have the same schema, and going direct to a `pa.Table` does not crash.

### OS:

x64

### DuckDB Version:

0.9.2

### DuckDB Client:

Python

### Full Name:

Matt Aubury

### Affiliation:

.

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a nightly build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by Tishj**
This PR fixes #10608 

This issue identified some issues surrounding lifetime management of ArrowArrays we are referencing to be able to zero-copy the data.

### Summary of changes

`arrow_owned_data` is gone from the `ArrowScanLocalState` class
Instead we now put a copy of the `shared_ptr<ArrowArrayWrapper> chunk` in the `ArrowArrayScanState`, this gets reset every time we fetch a new chunk.

`dictionary` was created once and was assumed to be valid for every subsequent fetched chunk.
We assign the ArrowAuxiliaryData to the created `dictionary`, just seems less error-prone that it owns the data it references.
~~Now we also reset the `dictionary` every time we fetch a new chunk.~~

Now we check the `arrow.dictionary` pointer against the pointer we used to create the cached Vector
If they do not match, we rebuild the dictionary Vector

The linked issue uncovered this issue because it combines multiple tables into one, each with their own dictionary.
So whenever we switch to a new "table", we need to reset our cached dictionary

### CODE REVIEW

File: src/function/table/arrow/arrow_array_scan_state.cpp [hunk 1]
```diff
@@ -1,29 +1,55 @@
...
 bool ArrowArrayScanState::HasDictionary() const {
 	return dictionary != nullptr;
 }
 
+bool ArrowArrayScanState::CacheOutdated(ArrowArray *dictionary) const {
+	if (!dictionary) {
+		// Not cached
+		return true;
+	}
+	if (dictionary == arrow_dictionary.get()) {
```
Comment (pdet) [inline src/function/table/arrow/arrow_array_scan_state.cpp:46]: I guess this should be fine, but maybe let's add a small regression test over the dictionary data chunks just the pointer comparison is not funky?

**Comment by Tishj** [comment]
I assume the data set we're creating in the regression test is too big for the runner
I'll scale it down

**Comment by Mytherin** [comment]
@pdet is this ready to merge?

**Review by pdet — APPROVED** [review]
lgtm

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**